include/wx/peninfo.h

```diff
@@ -0,0 +1,145 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        wx/peninfo.h
+// Purpose:     Header for wxPenInfo/wxGraphicsPenInfo
+// Author:      Adrien Tétar
+// Created:     2017-05-03
+// Copyright:   (c) The wxWidgets Project Developers
+// Licence:     wxWindows Licence
+/////////////////////////////////////////////////////////////////////////////
+
+#ifndef _WX_PENINFO_H_BASE_
+#define _WX_PENINFO_H_BASE_
+
+#include "wx/defs.h"
+
+template <class T>
+class wxPenInfoBase
+{
+public:
+    // Setters for the various attributes. All of them return the object itself
+    // so that the calls to them could be chained.
+
+    T& Colour(const wxColour& colour)
+        { m_colour = colour; return static_cast<T&>(*this); }
+
+    T& Style(wxPenStyle style)
+        { m_style = style; return static_cast<T&>(*this); }
+    T& Stipple(const wxBitmap& stipple)
+        { m_stipple = stipple; m_style = wxPENSTYLE_STIPPLE; return static_cast<T&>(*this); }
+    T& Dashes(int nb_dashes, const wxDash *dash)
+        { m_nb_dashes = nb_dashes; m_dash = (wxDash *)dash; return static_cast<T&>(*this); }
+    T& Join(wxPenJoin join)
+        { m_join = join; return static_cast<T&>(*this); }
+    T& Cap(wxPenCap cap)
+        { m_cap = cap; return static_cast<T&>(*this); }
+
+    // Accessors are mostly meant to be used by wxPen itself.
+
+    wxColour GetColour() const { return m_colour; }
+    wxBitmap GetStipple() const { return m_stipple; }
+    wxPenStyle GetStyle() const { return m_style; }
+    wxPenJoin GetJoin() const { return m_join; }
+    wxPenCap GetCap() const { return m_cap; }
+    int GetDashes(wxDash **ptr) const { *ptr = m_dash; return m_nb_dashes; }
+
+    // Convenience
+
+    bool IsTransparent() const { return m_style == wxPENSTYLE_TRANSPARENT; }
+
+private:
+    void Init()
+    {
+        /* m_stipple skipped */
+        m_nb_dashes = 0;
+        m_dash = NULL;
+        m_join = wxJOIN_ROUND;
+        m_cap = wxCAP_ROUND;
+    }
+
+    wxColour m_colour;
+    wxBitmap m_stipple;
+    wxPenStyle m_style;
+    wxPenJoin m_join;
+    wxPenCap m_cap;
+
+    int m_nb_dashes;
+    wxDash* m_dash;
+};
+
+// ----------------------------------------------------------------------------
+// wxPenInfo describes a wxPen
+// ----------------------------------------------------------------------------
+
+class wxPenInfo : public wxPenInfoBase<wxPenInfo>
+{
+    using wxPenInfoBase<wxPenInfo>::wxPenInfoBase;
```

Comment:
This doesn't actually change anything for the compiler (the ctor is still public because it's public in the base class), but I'd put this in the public section of this class too, to avoid any confusion.

But, actually, wait: this is not needed at all because there are no ctors in the base class now that `wxGraphicsPenInfo` doesn't accept int width any more. So either the ctor from `int width` should be moved to the base class or this line should just be removed completely.